controllers/imagerepository_controller.go

```diff
@@ -79,6 +83,9 @@ func (r *ImageRepositoryReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
 
 	log := r.Log.WithValues("controller", strings.ToLower(imagev1alpha1.ImageRepositoryKind), "request", req.NamespacedName)
 
+	// record rediness metric
+	defer r.recordReadinessMetric(imageRepo)
```

Comment:
This passes imageRepo by value, including the value of the status and the value of the conditions slice. This means the conditions examined by recordReadinessMetrics will be whatever is the case at this line, rather than whatever has been calculated by the end of `Reconcile`. To fix this, pass a pointer, or if you don't want to depart from the convention set by other controllers, use a thunk `func() { r.recordReadinessMetric(imageRepo) }` so the evaluation is delayed.

Reply:
Great observation, I will change it to pass it by reference and I think this pattern should be followed by other controllers too, it does seem to be a common thing that we always exit the reconciliation loop immediately after this call is made.

Reply:
There's a fair bit of stuff (methods like `event()` and `recordReadinessMetric()`, but maybe control flow too) that could be factored out into a GOTK controller shell. I may open a discussion about it, but if someone beat me to that I would not be upset :-)

Reply:
Yes, I've noticed this too while jumping from one repo to the other, quite a lot of common code block candidates eligible for extraction into `pkg` maybe.